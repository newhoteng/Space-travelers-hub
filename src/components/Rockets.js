import { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { getRockets, reserveRocket } from '../redux/Rockets/RocketsSlice';

function Rockets() {
  const dispatch = useDispatch();
  const { rocket, isLoading } = useSelector((state) => state.rocket);

  useEffect(() => {
    if (isLoading === false) dispatch(getRockets());
  }, [dispatch, isLoading]);

  return (
    <div className="rockets-container">
      {isLoading && rocket.map((rocket) => (
        <div key={rocket.id} className="container">
          <div className="article">
            <img src={rocket.image} alt={rocket.name} className="rocket-image" />
          </div>
          <div className="desc">
            <h2 className="rocket-title">{rocket.name}</h2>
            <p className="rocket-details">
              {rocket.reserved && <span className="reserved">Reserved</span>}
              {rocket.description}
            </p>
            <button
              type="button"
              className={rocket.reserved ? 'cancel' : 'reserve'}
              onClick={() => dispatch(reserveRocket(rocket.id))}
            >
              {rocket.reserved ? 'Cancel Reservation' : 'Reserve Rocket'}
            </button>
          </div>
        </div>
      ))}
    </div>
  );
}

export default Rockets;
