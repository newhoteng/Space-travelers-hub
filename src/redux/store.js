import { configureStore } from '@reduxjs/toolkit';
import missionsReducer from './missions/missionsSlice';
import dragonReducer from './Dragons/DragonsSlice';
import rocketReducer from './Rockets/RocketsSlice';

export const store = configureStore({
  reducer: {
    missions: missionsReducer,
    dragon: dragonReducer,
    rocket: rocketReducer,
  },
});

export default store;
