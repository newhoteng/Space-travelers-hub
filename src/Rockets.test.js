import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { store } from './redux/store';

// Component imports
import Rockets from './components/Rockets';

// Function imports
import reducer, { reserveRocket } from './redux/Rockets/RocketsSlice';

describe('Rockets Slice', () => {
  test('should handle reserving a rocket', () => {
    const state = {
      rocket: [
        {
          id: '0', name: 'Falcon 1', description: 'The best', image: 'linkto image',
        },
        {
          id: '1', name: 'Falcon 9', description: 'Second best', image: 'linkto image',
        },
        {
          id: '2', name: 'Falcon Heavy', description: 'Third best', image: 'linkto image',
        },
      ],
      isLoading: false,
    };

    expect(reducer(state, reserveRocket('1'))).toEqual(
      {
        rocket: [
          {
            id: '0', name: 'Falcon 1', description: 'The best', image: 'linkto image',
          },
          {
            id: '1', name: 'Falcon 9', description: 'Second best', image: 'linkto image', reserved: true,
          },
          {
            id: '2', name: 'Falcon Heavy', description: 'Third best', image: 'linkto image',
          },
        ],
        isLoading: false,
      },
    );
  });
});

it('rockets renders correctly', () => {
  const rockets = renderer
    .create(
      <Provider store={store}>
        <Rockets page="/">Rockets Page</Rockets>
      </Provider>,
    )
    .toJSON();
  expect(rockets).toMatchSnapshot();
});

it('renders Rockets component', () => {
  render(
    <Provider store={store}>
      <Rockets />
    </Provider>,
  );
});
