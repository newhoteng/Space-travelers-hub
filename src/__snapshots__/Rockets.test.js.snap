// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rockets renders correctly 1`] = `
<div
  className="rockets-container"
/>
`;
